# Applying Algorithm

In [29]:
from numpy import array
import pandas as pd

In [30]:
train_dataset=pd.read_csv('train_dataset.csv')
test_dataset=pd.read_csv('test_dataset.csv')

Python libraries imported to read train and test dataset csv files 

In [31]:
train_dataset.info()
train_dataset2=train_dataset.fillna(train_dataset.mean())
train_dataset2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16500 entries, 0 to 16499
Data columns (total 23 columns):
Unnamed: 0                16500 non-null int64
House ID                  16500 non-null float64
bathrooms                 15536 non-null float64
bedrooms                  15851 non-null float64
blessings                 16500 non-null float64
builder                   15352 non-null float64
capital_distance          15669 non-null float64
dateb                     16500 non-null float64
datep                     16500 non-null float64
dining_rooms              14862 non-null float64
dock_distance             14859 non-null float64
farm                      16500 non-null float64
garden                    14515 non-null float64
holy_tree                 13234 non-null float64
king_visited              16500 non-null float64
knights_house_distance    15669 non-null float64
location                  14829 non-null float64
renovation                14871 non-null float64
river_dista

Missing datas having Nan values were filled using mean in train dataset

In [32]:
#test_dataset1=test_dataset.set_index('House ID')
#test_dataset.interpolate(method='barycentric',inplace=True)
test_dataset.info()
test_dataset2=test_dataset.fillna(test_dataset.mean())
test_dataset2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3500 entries, 0 to 3499
Data columns (total 23 columns):
Unnamed: 0                3500 non-null int64
House ID                  3500 non-null float64
bathrooms                 3298 non-null float64
bedrooms                  3374 non-null float64
blessings                 3500 non-null float64
builder                   3255 non-null float64
capital_distance          3316 non-null float64
dateb                     3500 non-null float64
datep                     3500 non-null float64
dining_rooms              3148 non-null float64
dock_distance             3121 non-null float64
farm                      3500 non-null float64
garden                    3071 non-null float64
holy_tree                 2796 non-null float64
king_visited              3500 non-null float64
knights_house_distance    3316 non-null float64
location                  3134 non-null float64
renovation                3136 non-null float64
river_distance            3500 

Missing datas having Nan values were filled using mean in test dataset

In [33]:
train_dataset2.head()

Unnamed: 0      House ID  bathrooms  bedrooms  blessings   builder  \
0           0  1.848823e+09   3.007209  4.000000      160.0  0.000000   
1           1  1.848823e+09   3.007209  4.000000      133.0  8.000000   
2           2  1.848823e+09   2.000000  3.001514       24.0  4.003192   
3           3  1.848824e+09   2.000000  2.000000      140.0  4.000000   
4           4  1.848824e+09   2.000000  2.000000      159.0  3.000000   

   capital_distance     dateb     datep  dining_rooms      ...        \
0          0.256239  584875.0  587511.0           3.0      ...         
1         26.882222  584271.0  588537.0           2.0      ...         
2        136.448973  584160.0  587116.0           3.0      ...         
3         19.885265  585167.0  588206.0           2.0      ...         
4          0.519507  585131.0  587968.0           3.0      ...         

   holy_tree  king_visited  knights_house_distance  location  renovation  \
0   1.000000           0.0               10.505662  3.000000         0.0   
1   0.870107           1.0               37.249456  0.000000         0.0   
2   1.000000           0.0              146.553349  0.993998         0.0   
3   0.870107           0.0               29.921162  0.000000         0.0   
4   1.000000           0.0               10.855780  0.000000         0.0   

   river_distance  royal_market  sorcerer_cursed  tower_distance  \
0       63.109038     69.486467              0.0       14.394903   
1       53.249336     34.145238              0.0      101.052598   
2       72.245533     78.445508              0.0       88.480988   
3        8.059939     92.698917              0.0       22.395136   
4       35.706541      3.605055              0.0       81.449646   

   Golden Grains  
0      1470912.0  
1      1241030.0  
2       838723.0  
3      1267804.0  
4      1497166.0  

[5 rows x 23 columns]

In [34]:
features=['dateb','datep','garden','dock_distance','royal_market','capital_distance','tower_distance','builder','river_distance','dining_rooms','bedrooms','bathrooms','king_visited','renovation','blessings','farm','location','sorcerer_cursed','knights_house_distance','holy_tree']
features1=['dateb','datep','dock_distance','builder','royal_market','capital_distance','tower_distance','river_distance','dining_rooms','bedrooms','bathrooms','king_visited','blessings','farm','location','sorcerer_cursed','knights_house_distance','holy_tree']

'features' contains all the features of the train dataset and 'features1' contains all the festures excluding those we dropped
for eg-garden,houseId,builder,renovation.

In [35]:
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LinearRegression
from time import time

traind_x=train_dataset2[features1]
traind_y=train_dataset2['Golden Grains']
t0=time()
model=LinearRegression(normalize=False)
scores=cross_val_score(model,traind_x,traind_y,cv=5,scoring='r2')
print('mean score is',scores.mean())
print('time by LR is ',round(time()-t0,2))

mean score is 0.943112614318
time by LR is  0.18


Applying Linear Regression with cross validation mean score of 94.311 was obtained within 0.18 seconds time

In [36]:
from sklearn.ensemble import GradientBoostingRegressor
t0=time()
model1=GradientBoostingRegressor()
scores=cross_val_score(model1,traind_x,traind_y,cv=5,scoring='r2')
print('mean score is',scores.mean())
print('time by GBR is ',round(time()-t0,2))

mean score is 0.968622305065
time by GBR is  17.13


Applying Gradient Boosting with cross validation without parameter tuning mean score of 96.862 was obtained with 17.13 seconds time
It can be observed that this algorithm is taking more time but is increasing the accuracy

In [37]:
from sklearn.ensemble import GradientBoostingRegressor
t0=time()
model2=GradientBoostingRegressor(n_estimators=300)
scores=cross_val_score(model2,traind_x,traind_y,cv=5,scoring='r2')
print('mean score is',scores.mean())
print('time by GBR is ',round(time()-t0,2))

mean score is 0.994703235998
time by GBR is  47.78


We can observe that by tuning parameters mean score increased

In [38]:
t0=time()
model3=GradientBoostingRegressor(n_estimators=350)
scores=cross_val_score(model3,traind_x,traind_y,cv=5,scoring='r2')
print('mean score is',scores.mean())
print('time by GBR is ',round(time()-t0,2))

mean score is 0.995416091841
time by GBR is  54.58


Here the change in mean score by increasing the n_estimators is less and also more number of n_estimators is increasing our time 
and hence its better to keep n_estimators to 300 and test this for our final answer

In [39]:
from sklearn import metrics
testd_x=test_dataset2[features1]
model2.fit(traind_x,traind_y)
prediction=model2.predict(testd_x)
sol=pd.read_csv('solution.csv')
sol.drop('House ID',axis=1,inplace=True)
score=metrics.r2_score(prediction,sol)
print(score)

0.991647240201


fitting Gradient Boosting with n_estimators 300 and predicting for test data and comparing with answer we get final score

In [40]:
raw_data={'Golden_grains' : prediction}
df = pd.DataFrame(raw_data, columns = ['Golden_grains'])
df.to_csv('housingprices_ans.csv')

We finally obtain the housingprices_ans as our solution in csv format